include/envoy/event/dispatcher.h

```diff
@@ -260,6 +260,12 @@ class Dispatcher : public DispatcherBase {
    */
   virtual void post(PostCb callback) PURE;
 
+  /**
+   * Similar to `post()` but will destroy passed callback. This simulates posting move only
+   * function.
+   */
+  virtual void movePost(PostCb&& callback) PURE;
```

Comment:
We should consider having post take an r-value reference instead of adding a new API method.

Reply:
Let's first make it work and then make it perfect.
This movePost we must do copy then immediately destroy the passed function, which is even stricter than std::move(). A new name is the best I can warn the user the pass-in callback will be mutated.

Reply:
When do you expect to clean up the Dispatcher API?

Reply:
The latest commit 
1. change the legacy `post(PostCb callback)` to `post(const PostCb& callback)`. It has the same side effect. The value passed is copied.
2. rename movePost(PostCb&& callback) to post(PostCb&& callback). So it's a overload. The caller can pass in a r-value functor and post() will set nullptr to the passed in r-value reference.

IMHO this approach should introduce no risk given post(PostCb&&) erase the passed in functor.

Reply:
The overload breaks lots of test:
1. `post(_)` doesn't work because compiler cannot detect which post method should be using
2. The existing code might switch to `post(Cb&&)`. Not bad but I need to fix them.

I can fix the test cases in Envoy code base, but this change will break test cases in the project on top of envoy code. 

Do you think I should go further on the overload path? Though I think it has the least damage out of mock code

Reply:
outdated: movePost renamed to deleteInDispatcherThread and pass unique ptr there